clarity/src/vm/ast/traits_resolver/mod.rs

```diff
@@ -63,26 +64,30 @@ impl TraitsResolver {
                     match (&args[0].pre_expr, &args[1].pre_expr) {
                         (Atom(trait_name), List(trait_definition)) => {
                             // Check for collisions
-                            if contract_ast.referenced_traits.contains_key(trait_name) {
-                                return Err(
-                                    ParseErrors::NameAlreadyUsed(trait_name.to_string()).into()
-                                );
+                            match contract_ast.referenced_traits.entry(trait_name.to_owned()) {
+                                Entry::Occupied(_) => {
+                                    return Err(ParseErrors::NameAlreadyUsed(
+                                        trait_name.to_string(),
+                                    )
+                                    .into());
+                                }
+                                Entry::Vacant(e) => {
+                                    // Traverse and probe for generics nested in the trait definition
+                                    self.probe_for_generics(
+                                        trait_definition.iter(),
+                                        &mut referenced_traits,
+                                        true,
+                                    )?;
+
+                                    let trait_id = TraitIdentifier {
+                                        name: trait_name.clone(),
+                                        contract_identifier: contract_ast
+                                            .contract_identifier
+                                            .clone(),
+                                    };
+                                    e.insert(TraitDefinition::Defined(trait_id));
+                                }
                             }
```

Comment:
One minor nitpick here: We should be consistent with Rust naming conventions and use
 - `map` for infallible conversions
 - `and_then` for fallible conversions which return a `Result`

Otherwise this looks cool and would be a nice addition to the `StacksHashMap` type

Reply:
I'll finish the stacking testing and do into this afterwards

Reply:
@jbencin recommend `occupied_and_then()` + `vacant_and_then()` in that case

Reply:
@ASuciuX here comes an update for the `move` issue you ran into:

**Trait + Impl**
```rust
pub trait EntryExt<'a, K: 'a, V: 'a, S: 'a>
where
    S: BuildHasher,
    Self: Sized,
{
    fn occupied_and_then<T, F, E>(self, f: F) -> Result<Self, E>
    where
        F: FnOnce(&OccupiedEntry<'a, K, V, S>) -> Result<T, E>;

    fn or_insert_fallible<F, E>(self, default: F) -> Result<&'a mut V, E>
    where
        F: FnOnce() -> Result<V, E>;

    fn vacant_or<E>(self, err: E) -> Result<Self, E>;
}

impl<'a, K: 'a, V: 'a, S: 'a> EntryExt<'a, K, V, S> for Entry<'a, K, V, S>
where
    S: BuildHasher,
    K: Hash
{    
    fn occupied_and_then<T, F, E>(self, f: F) -> Result<Self, E>
    where 
        F: FnOnce(&OccupiedEntry<'a, K, V, S>) -> Result<T, E>
    {
        match self {
            Entry::Occupied(ref e) => {
                f(e)?;
                Ok(self)
            },
            Entry::Vacant(_) => Ok(self)
        }
    }

    fn or_insert_fallible<F, E>(self, default: F) -> Result<&'a mut V, E>
    where
        F: FnOnce() -> Result<V, E>
    {
        match self {
            Entry::Occupied(e) => Ok(e.into_mut()),
            Entry::Vacant(e) => Ok(e.insert(default()?)),
        }
    }
    
    fn vacant_or<E>(self, err: E) -> Result<Self, E> {
        match self {
            Entry::Occupied(_) => Err(err),
            Entry::Vacant(_) => Ok(self)
        }
    }
}
```

**Usage**
```rust
match (&args[0].pre_expr, &args[1].pre_expr) {
    (Atom(trait_name), List(trait_definition)) => {
        contract_ast
            .referenced_traits
            .entry(trait_name.to_owned())
            .vacant_or(ParseErrors::NameAlreadyUsed(trait_name.to_string()))?
            .or_insert_fallible(|| {
                self.probe_for_generics(
                    trait_definition.iter(),
                    &mut referenced_traits,
                    true,
                ).map_err(|e| e.err)?;

                let trait_id = TraitIdentifier {
                    name: trait_name.clone(),
                    contract_identifier: contract_ast
                        .contract_identifier
                        .clone(),
                };
                Ok::<_, ParseErrors>(TraitDefinition::Defined(trait_id))
            })?;
    }
    _ => return Err(ParseErrors::DefineTraitBadSignature.into()),
}
```

Reply:
@jbencin also updated naming in the above. couldn't think of anything awesome for `or_insert_fallible`... `or_insert` etc. are already taken by hashbrown itself...

Reply:
`or_try_insert` perhaps

Reply:
@jcnelson I invite you to give your input on the above as well as you had comments on the original impl.